src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2883,41 +2884,42 @@ def validate_encryption_data(self, job, verrors, encryption_dict, schema):
 
         key_format = ZFSKeyFormat(encryption_dict['key_format'])
         passphrase_key_format = key_format == ZFSKeyFormat.PASSPHRASE
-        if not encryption_dict['key_file'] and not passphrase_key_format and not encryption_dict['generate_key']:
-            verrors.add(
-                f'{schema}.key_format',
-                'Please provide a key file or select generate_key to automatically generate '
-                'a key when key_format is not "PASSPHRASE"'
-            )
+        key = encryption_dict['key']
+        passphrase = encryption_dict['passphrase']
 
         if passphrase_key_format:
             if encryption_dict['generate_key']:
                 verrors.add(
                     f'{schema}.generate_key',
                     'Must be disabled when key format is set to "PASSPHRASE".'
                 )
-            elif not encryption_dict['passphrase']:
+            elif not passphrase:
                 verrors.add(f'{schema}.passphrase', 'Passphrase is required when key format is "PASSPHRASE"')
-        elif encryption_dict['passphrase']:
-            verrors.add(
-                f'{schema}.passphrase',
-                'Should not be provided when key format is not "PASSPHRASE"'
-            )
+        else:
+            if passphrase:
+                verrors.add(f'{schema}.passphrase', 'Must not be specified when key format is not "PASSPHRASE"')
+            if not key and not encryption_dict['key_file'] and not encryption_dict['generate_key']:
+                verrors.add(
+                    f'{schema}.key_format',
+                    'Please provide a key, key file or select generate_key to automatically generate '
+                    'a key when key_format is not "PASSPHRASE"'
+                )
+            if key and encryption_dict['generate_key']:
+                verrors.add(f'{schema}.key', 'Must not be specified when automatic generation of key is desired')
 
         if not verrors:
-            key = encryption_dict['passphrase']
+            key = key or passphrase
             if encryption_dict['generate_key']:
-                key = getattr(secrets, f'token_{"hex" if key_format == ZFSKeyFormat.HEX else "bytes"}')(32)
+                key = secrets.token_hex(32)
             elif not key:
                 job.check_pipe('input')
-                key = job.pipes.input.r.read(32 if key_format == ZFSKeyFormat.RAW else 64)
+                key = job.pipes.input.r.read(64)
                 # We would like to ensure key matches specified key format
-                if key_format == ZFSKeyFormat.HEX:
-                    try:
-                        key.decode()
-                    except UnicodeDecodeError:
-                        verrors.add(f'{schema}.key_file', 'Unable to decode "HEX" key')
-                        return {}
+                try:
+                    key = key.decode()
```

Comment:
only .decode() seems too weak, it will decode utf8 chars.

Should we probably do .decode('ascii') and/or perhaps make sure its hexadecimal?